### ISSUE TITLE
KMSAccessDeniedException When Attempting Invoke

### ISSUE THREAD

**Issue by chrisforrette** [apex/apex#645]
I'm using Apex and Terraform and referenced the [babel-webpack](https://github.com/apex/apex/tree/master/_examples/babel-webpack) example in the `_examples` folder to configure a Babel/Webpack compile step into the build. After I've deployed and try to invoke my Lambda function using the Node AWS SDK, I get the following exception:

`KMSAccessDeniedException: Lambda was unable to decrypt the environment variables because KMS access was denied. Please check the function's KMS key settings. KMS Exception: AccessDeniedExceptionKMS Message: The ciphertext refers to a customer master key that does not exist, does not exist in this region, or you are not allowed to access.`

As far as I can tell, you can't configure the KMS key being set with Apex, and when I look under the `Configuration > Advanced Settings` tab for my function in the AWS console, I see `(default) aws/lambda` for the "KMS Key", which I assume was created by default, and which I would assume would not have restricted access to it.

The thing that makes this particularly confusing is that this doesn't happen consistently. I'm namespacing my functions by prefixing an environment to the name, and using a "testing-*" version I wasn't experiencing this issue. At some point, a deploy to the environment I'm having the issue in now started working after I removed a Babel plugin being used (`transform-runtime`), but deployed again later and it returned.

Is there a way to configure KMS keys used by functions in Apex? And/or do I need to grant permission to the KMS key (even the default one..?) to the role being assigned to each function? Or is there something else I'm missing? My problem might just be my lack of understanding of how KMS and Lambda play together—it seems like it's implicit, but apparently wants something explicit from me. Before I dive in and read the entire KMS docs, I'm hoping I'm just missing something (not) obvious (to me).

  - __Operating system__: Mac OS 10.11.6
  - __Apex Version__: 0.11.0

Inline IAM policy attached to the Lambda role:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": "logs:CreateLogGroup",
            "Resource": "arn:aws:logs:us-west-2:(account id):*"
        },
        {
            "Effect": "Allow",
            "Action": [
                "logs:CreateLogStream",
                "logs:PutLogEvents"
            ],
            "Resource": [
                "arn:aws:logs:us-west-2:(account id):*"
            ]
        },
        {
            "Effect": "Allow",
            "Action": [
                "s3:ListBucket"
            ],
            "Resource": [
                "arn:aws:s3:::a-bucket-lambda-needs-access-to"
            ]
        },
        {
            "Effect": "Allow",
            "Action": [
                "s3:PutObject",
                "s3:GetObject",
                "s3:DeleteObject"
            ],
            "Resource": [
                "arn:aws:s3:::a-bucket-lambda-needs-access-to/*"
            ]
        }
    ]
}
```

**Comment by tj** [comment]
Hmm I haven't had permission issues with the default KMS stuff for Lambda, can't comment on non-defaults since I haven't bothered to try that

**Comment by chrisforrette** [comment]
@tj Thanks for your response.

I managed to get it working by manually setting the `KMSKeyArn` property to the ARN of the default KMS key using the [`updateFunctionConfiguration` function on the Node client](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Lambda.html#updateFunctionConfiguration-property). It was previously set to `null` and I would think it would pick up the default KMS key in that case, so I'm a little confused by that.

In the case of Apex, is there a way to set the `KMSKeyArn` property at the project or function level?

**Comment by tj** [comment]
Not at the moment, what's the use-case for changing the default? I'm not against it but I can't see myself using it.

**Comment by chrisforrette** [comment]
@tj The use case is obviously entirely selfish and to appease the requirement in my case to set the KMS ARN. This is kind of unfortunate for my automated deployment process—I suppose I'll need to write a script to run after deploys to manually set the KMS ARNs for my functions if the issue persists.

In general, having such control would be handy if you're particularly concerned with security stuff, and I would think it optimal to pass through all or most of the available configuration options for a function.

In terms of Apex, I'm unsure of what the goals of the project are, but it sounds like it's mainly serving your personal work flow and you happen to be (generously) sharing it. I wish I could offer to open a PR for it, but I haven't written a line of Go beyond "Hello World".

Anyway—thanks for everything!

**Comment by tj** [comment]
Yep, I try to stick to stuff I use personally, but if it's not a huge change or maintenance/QA burden I totally don't mind adding stuff that other people need. Happy to check out a PR :D

**Comment by DaveBlooman** [comment]
#648 :)
